logsmanagement/circular_buffer.c

```diff
@@ -0,0 +1,264 @@
+        return 0;
+    }
+    
+    if(required_space > buff->in->data_max_size) {
+        buff->in->data_max_size = required_space;
+        buff->in->data = reallocz(buff->in->data, buff->in->data_max_size);
+    }
+
+    buff->in->text_compressed_size = requested_text_compressed_space;
+
+    return requested_text_space;
+}
+
+int circ_buff_insert(Circ_buff_t *const buff){
+    
+    if (unlikely((buff->head % buff->num_of_items == buff->tail % buff->num_of_items) && buff->full)) {
+        error("Logs circular buffer out of space! Losing data!");
+        m_assert(0, "Buff full");
+        // TODO: How to handle this case, when circular buffer is out of space?
+        return -1;
+    }
+
+    Circ_buff_item_t *cur_item = &buff->items[buff->head % buff->num_of_items];
+
+    char *tmp_data = cur_item->data;
+    size_t tmp_data_max_size = cur_item->data_max_size;
+
+    cur_item->status = buff->in->status;
+    cur_item->timestamp = buff->in->timestamp;
+    cur_item->data = buff->in->data;
+    cur_item->text_size = buff->in->text_size;
+    cur_item->text_compressed = buff->in->text_compressed;
+    cur_item->text_compressed_size = buff->in->text_compressed_size;
+    cur_item->data_max_size = buff->in->data_max_size;
+    
+    buff->in->status = CIRC_BUFF_ITEM_STATUS_UNPROCESSED;
+    buff->in->timestamp = 0;
+    buff->in->data = tmp_data;
+    buff->in->text_size = 0;
+    // buff->in->text_compressed = tmp_data;
+    buff->in->text_compressed_size = 0;
+    buff->in->data_max_size = tmp_data_max_size;
+
+    buff->text_size_total += cur_item->text_size;
+    buff->text_compressed_size_total += cur_item->text_compressed_size;
+    buff->compression_ratio = buff->text_compressed_size_total ? buff->text_size_total / buff->text_compressed_size_total : 0;
+
+    buff->head++;
+    if(unlikely(buff->head % buff->num_of_items == buff->tail % buff->num_of_items)) buff->full = 1;
+
+    return 0;
+}
+
+Circ_buff_item_t *circ_buff_read_item(Circ_buff_t *const buff) {
+
+    Circ_buff_item_t *item = &buff->items[buff->read % buff->num_of_items];
+
+    m_assert(item->status >= 0 && item->status <= 3, "Invalid status");
+
+    if( buff->read % buff->num_of_items == buff->head % buff->num_of_items || /* No more records to be retrieved in the buffer */
+        item->status != CIRC_BUFF_ITEM_STATUS_DONE /* current item either not parsed or streamed */){
+        
+        buff->tail = buff->read;
+        /* Tail moved so update buff full flag in case it is set */
+        buff->full = 0;
+        buff->text_size_total = 0;
+        buff->text_compressed_size_total = 0;
+        return NULL;
+    }
+
+    buff->read++;
+
+    return item;
+}
+
+/**
+ * @brief Create a new circular buffer.
+ * @param num_of_items Number of Circ_buff_item_t items in the buffer.
+ * @param max_size Maximum memory the circular buffer can occupy.
+ * @return Pointer to the new circular buffer structure.
+ */
```

Comment:
I loved the fact you described arguments here, but you did not write for the previous functions, I suggest you use the same pattern for all functions. :)

Reply:
Done!